File: backend/coding_intelligence.py
"""
Coding validation, modifier suggestions, and pre-submission denial risk (rule-based, <1s typical).
Not a substitute for certified coding — assists intake and appeal prep.
"""
from __future__ import annotations

import re
from typing import Any, Dict, List, Optional

# High-level E/M ranges (simplified)
_EM_RANGE = re.compile(r"^99[2-4]\d{2}$")
_HIGH_OFFICE_EM = re.compile(r"^992(0[4-5]|1[3-5])$")
_PROCEDURE_SURGICAL = re.compile(r"^(1\d{4}|2\d{4}|3\d{4}|4\d{4}|5\d{4}|6\d{4})$")
_IMAGING = re.compile(r"^(7\d{4})$")


def _split_codes(raw: Optional[Any]) -> List[str]:
    if raw is None:
        return []
    if isinstance(raw, list):
        parts = raw
    else:
        parts = re.split(r"[\s,;]+", str(raw).strip())
    out: List[str] = []
    for p in parts:
        p = p.strip().upper().replace(".", "")
        if not p:
            continue
        base = re.match(r"^(\d{5})", p)
        if base:
            out.append(base.group(1))
        elif re.match(r"^\d{5}$", p):
            out.append(p)
    return list(dict.fromkeys(out))[:40]


def _split_icd(raw: Optional[Any]) -> List[str]:
    if raw is None:
        return []
    if isinstance(raw, list):
        parts = raw
    else:
        parts = re.split(r"[\s,;]+", str(raw).strip())
    out: List[str] = []
    for p in parts:
        p = re.sub(r"[^A-Z0-9.]", "", p.upper())
        if len(p) >= 3:
            out.append(p)
    return list(dict.fromkeys(out))[:24]


def _icd_specificity(icd: str) -> str:
    core = icd.replace(".", "")
    if len(core) <= 3:
        return "unspecified"
    if core.startswith("Z") and len(core) <= 5:
        return "general_z"
    if re.match(r"^[A-Z]\d{2}\d{3}$", core) or re.match(r"^[A-Z]\d{2}\.\d", icd):
        return "specific"
    return "moderate"


def _normalize_validate_pair(cpt_codes: Any, icd_codes: Any, kwargs: Any) -> tuple[Any, Any]:
    if isinstance(cpt_codes, dict):
        d = cpt_codes
        return (
            d.get("cptCodes") or d.get("cpt_codes"),
            d.get("icdCodes") or d.get("icd10_codes") or d.get("icd_codes"),
        )
    cpt = kwargs.get("cptCodes") if kwargs.get("cptCodes") is not None else cpt_codes
    icd = kwargs.get("icdCodes") if kwargs.get("icdCodes") is not None else icd_codes
    return (cpt, icd)


def validateCoding(cpt_codes: Any = None, icd_codes: Any = None, **kwargs: Any) -> Dict[str, Any]:
    """
    validateCoding({ cptCodes, icdCodes }) or validateCoding(cpt_str, icd_str)
    Returns valid, issues, suggestions, weakIcdCodes for UI highlight.
    """
    cpt_in, icd_in = _normalize_validate_pair(cpt_codes, icd_codes, kwargs)
    cpts = _split_codes(cpt_in)
    icds = _split_icd(icd_in)
    issues: List[str] = []
    suggestions: List[str] = []
    weak_icd: List[str] = []

    if not cpts:
        issues.append("No CPT/HCPCS codes provided — cannot validate medical necessity pairing.")
        suggestions.append("Enter at least one procedure or visit code.")

    if not icds:
        issues.append("No ICD-10 diagnosis codes provided — missing diagnosis linkage.")
        suggestions.append("Add diagnosis codes that support medical necessity for each service line.")

    em_codes = [c for c in cpts if _EM_RANGE.match(c)]
    proc_codes = [c for c in cpts if _PROCEDURE_SURGICAL.match(c)]
    img_codes = [c for c in cpts if _IMAGING.match(c)]

    for icd in icds:
        spec = _icd_specificity(icd)
        if spec in ("unspecified", "general_z"):
            weak_icd.append(icd)

    for em in em_codes:
        if icds:
            only_z = all(i.startswith("Z") for i in icds)
            if only_z and _HIGH_OFFICE_EM.match(em):
                issues.append(f"ICD-10 pattern may not support CPT {em} (high-level E/M with only Z-series diagnoses).")
                suggestions.append("Add active diagnoses documented for this visit or align E/M level with payer policy.")
            if only_z and em >= "99204" and not _HIGH_OFFICE_EM.match(em):
                issues.append(f"ICD-10 may not fully support CPT {em} for medical necessity if only screening codes apply.")
                suggestions.append("Ensure visit diagnosis matches the work performed.")

    if em_codes and icds:
        only_z_screening = all(i.startswith("Z") for i in icds)
        if only_z_screening and any(c >= "99204" for c in em_codes):
            if not any("ICD-10 pattern may not support CPT" in x for x in issues):
                issues.append("Diagnosis is too general or screening-only for the service complexity reported.")
                suggestions.append("Use diagnoses that support medical necessity for the E/M level billed.")

    if em_codes and proc_codes:
        issues.append("E/M and procedure on the same encounter often require modifier -25 on the E/M when distinct.")
        suggestions.append("Confirm separate E/M service from procedure; consider modifier -25 if payer rules allow.")

    for icd in icds:
        if _icd_specificity(icd) == "unspecified":
            issues.append(f"ICD-10 {icd} appears truncated or unspecified — may lack specificity for the service.")
            suggestions.append("Use the most specific billable ICD-10 supported by documentation.")

    if proc_codes and not em_codes and len(icds) == 1 and _icd_specificity(icds[0]) in ("general_z", "unspecified"):
        issues.append("Procedure coding with a single broad or Z-code diagnosis may indicate incomplete linkage.")
        suggestions.append("Link complications, anatomy, or active conditions as documented.")

    if img_codes and icds:
        if all(_icd_specificity(i) in ("unspecified", "general_z", "moderate") for i in icds):
            issues.append("Imaging CPT with broadly stated diagnoses — payer may question medical necessity specificity.")
            suggestions.append("Tighten ICD-10 to findings or symptoms that justify the study.")

    if img_codes:
        issues.append("Imaging CPT codes often require prior authorization by payer and policy.")
        suggestions.append("Verify authorization on file for the DOS and modality.")

    valid = len(issues) == 0
    return {
        "valid": valid,
        "issues": issues[:14],
        "suggestions": suggestions[:14],
        "weakIcdCodes": list(dict.fromkeys(weak_icd))[:24],
        "cpt_parsed": cpts,
        "icd_parsed": icds,
    }


def _parse_carc_from_text(blob: str) -> List[int]:
    out: List[int] = []
    for m in re.finditer(r"(?:CO|PR|PI|OA)?[-\s]?(\d{1,3})", (blob or "").upper()):
        try:
            n = int(m.group(1))
            if 0 < n <= 999:
                out.append(n)
        except ValueError:
            pass
    return list(dict.fromkeys(out))[:24]


def _carc_from_any(carc_codes: Any, denial_blob: str) -> List[int]:
    nums = _parse_carc_from_text(denial_blob)
    if isinstance(carc_codes, list):
        for x in carc_codes:
            nums.extend(_parse_carc_from_text(str(x)))
    elif carc_codes:
        nums.extend(_parse_carc_from_text(str(carc_codes)))
    return list(dict.fromkeys(nums))[:24]


def _normalize_suggest_args(
    cpt_codes: Any, denial_codes: Any, context: Optional[Dict[str, Any]]
) -> tuple[List[str], List[int], Dict[str, Any], str]:
    ctx: Dict[str, Any] = dict(context or {})
    carcs: List[int]

    if isinstance(cpt_codes, dict):
        d = cpt_codes
        cpt_raw = d.get("cptCodes") or d.get("cpt_codes")
        carc_raw = d.get("carcCodes") or d.get("carc_codes")
        ctx = {**ctx, **(d.get("context") or {})}
        cpts = _split_codes(cpt_raw)
        blob = " ".join(str(x) for x in carc_raw) if isinstance(carc_raw, list) else str(carc_raw or "")
        carcs = _carc_from_any(carc_raw, blob)
        if denial_codes:
            extra = " ".join(str(x) for x in denial_codes) if isinstance(denial_codes, list) else str(denial_codes)
            blob = (blob + " " + extra).strip()
    else:
        cpts = _split_codes(cpt_codes)
        if isinstance(denial_codes, list):
            blob = " ".join(str(x) for x in denial_codes)
        else:
            blob = str(denial_codes or "")
        carcs = _carc_from_any(None, blob)

    blob = (blob + " " + str(ctx.get("denial_reason") or "")).strip()
    carcs = list(dict.fromkeys(carcs + _parse_carc_from_text(blob)))[:24]
    return cpts, carcs, ctx, blob


def suggestModifiers(
    cpt_codes: Any = None,
    denial_codes: Any = None,
    context: Optional[Dict[str, Any]] = None,
    **kwargs: Any,
) -> Dict[str, Any]:
    """
    suggestModifiers({ cptCodes, carcCodes, context }) or legacy (cpt, denial_blob, context).
    Returns recommendedModifiers and reasoning (one string per rule line).
    """
    cpts, carcs, ctx, blob = _normalize_suggest_args(cpt_codes, denial_codes, context)

    recommended: List[str] = []
    reasoning_lines: List[str] = []

    em_codes = [c for c in cpts if _EM_RANGE.match(c)]
    proc_codes = [c for c in cpts if _PROCEDURE_SURGICAL.match(c)]
    img_codes = [c for c in cpts if _IMAGING.match(c)]
    proc_like = proc_codes + img_codes

    if em_codes and proc_like:
        recommended.append("-25")
        reasoning_lines.append("Separate E/M service performed with procedure — distinct service when documented.")

    if 97 in carcs or "BUNDL" in blob.upper():
        if "-59" not in recommended and "-XS" not in recommended:
            recommended.append("-59")
        reasoning_lines.append("CARC 97 (bundling / inclusive payment) — distinct procedural service may require -59 or -XS.")

    if 4 in carcs or "MODIFIER" in blob.upper():
        if "-59" not in recommended:
            recommended.append("-59")
        reasoning_lines.append("CARC 4 or modifier-related denial — review NCCI and append -59 or appropriate modifier when distinct.")

    if ctx.get("global_period_unrelated") or "GLOBAL" in blob.upper():
        recommended.append("-24")
        reasoning_lines.append("Unrelated E/M during postoperative global period — use -24 when policy criteria are met.")

    if "REPEAT" in blob.upper() or 76 in carcs:
        recommended.append("-76")
        reasoning_lines.append("Repeat procedure by same physician — -76 when the same procedure is repeated.")

    if "DIFFERENT" in blob.upper() and "PHYSICIAN" in blob.upper():
        recommended.append("-77")
        reasoning_lines.append("Repeat procedure, different physician — consider -77.")

    n_proc = len([c for c in cpts if _PROCEDURE_SURGICAL.match(c) or _IMAGING.match(c)])
    if n_proc >= 2:
        if "-59" not in recommended and "-XS" not in recommended:
            recommended.append("-59")
        reasoning_lines.append("Multiple procedures — evaluate distinct services; -59 or -XS when separately reportable.")

    seen = set()
    mods: List[str] = []
    for m in recommended:
        if m not in seen:
            seen.add(m)
            mods.append(m)

    if not reasoning_lines:
        reasoning_lines.append("No automatic modifier rules triggered — verify payer policy and NCCI edits.")

    return {
        "recommendedModifiers": mods[:8],
        "reasoning": reasoning_lines[:12],
        "matched_carcs": carcs,
    }


def detectDenialRisk(input_data: Dict[str, Any]) -> Dict[str, Any]:
    """
    detectDenialRisk(input) — uses cptCodes, icdCodes, modifiers, carcCodes, payer, planType.
    """
    risks: List[str] = []
    recs: List[str] = []
    score = 0

    cpts = _split_codes(input_data.get("cpt_codes") or input_data.get("cptCodes"))
    icds = _split_icd(
        input_data.get("icd10_codes")
        or input_data.get("icd_codes")
        or input_data.get("diagnosis_code")
        or input_data.get("icdCodes")
    )
    mods = str(input_data.get("modifiers") or "").upper()
    payer = str(input_data.get("payer") or "")
    plan = str(input_data.get("planType") or input_data.get("plan_type") or "")

    carc_blob = " ".join(
        [
            str(input_data.get("denial_code") or ""),
            str(input_data.get("carc_codes") or ""),
            str(input_data.get("carcCodes") or ""),
        ]
    )
    if isinstance(input_data.get("carcCodes"), list):
        carc_blob += " " + " ".join(str(x) for x in input_data["carcCodes"])
    denial_blob = " ".join(
        [
            carc_blob,
            str(input_data.get("denial_reason") or ""),
            str(input_data.get("rarc_codes") or ""),
        ]
    )
    carc_nums = _parse_carc_from_text(denial_blob)

    if not cpts:
        score += 15
        risks.append("Missing procedure codes — high scrubber / medical necessity exposure.")
        recs.append("Enter all line-item CPT/HCPCS codes.")

    if not icds:
        score += 22
        risks.append("Missing diagnosis codes — high likelihood of medical necessity denial.")
        recs.append("Add ICD-10 codes tied to documentation.")

    for icd in icds:
        if _icd_specificity(icd) in ("unspecified", "general_z") and cpts:
            score += 12
            risks.append("Weak ICD-10 specificity relative to billed services.")
            recs.append("Use specific, billable diagnoses supported by the record.")
            break

    em_codes = [c for c in cpts if _EM_RANGE.match(c)]
    proc_codes = [c for c in cpts if _PROCEDURE_SURGICAL.match(c)]
    if em_codes and proc_codes and "-25" not in mods and "25" not in mods:
        score += 18
        risks.append("HIGH RISK: E/M plus procedure same day without modifier -25 — likely bundling or denial.")
        recs.append("Add modifier -25 to the E/M when separately identifiable.")

    if 197 in carc_nums or 16 in carc_nums:
        score += 14
        risks.append("Authorization / precertification cited on claim (CARC pattern).")
        recs.append("Secure authorization or retro-auth documentation before resubmission.")

    if 252 in _parse_carc_from_text(denial_blob):
        score += 8
        risks.append("Additional documentation or authorization gap (CARC 252 pattern).")
        recs.append("Submit complete clinical packet or auth proof.")

    if any(_IMAGING.match(c) for c in cpts):
        score += 8
        risks.append("Authorization-sensitive CPT: advanced imaging often requires precert.")
        recs.append("Confirm precertification and medical necessity narrative.")

    if plan.upper() in ("MEDICAID", "MEDICARE"):
        score += 5
        risks.append(f"{plan} plans often apply strict LCD / NCD medical necessity rules.")
        recs.append("Align diagnoses and modifiers with program coverage policies.")

    if "UNITED" in payer.upper() or "UHC" in payer.upper():
        score += 6
        risks.append("Known payer pattern: strict policy-driven medical necessity reviews.")
        recs.append("Cite applicable clinical policy in documentation.")

    if em_codes and icds and all(i.startswith("Z") for i in icds) and any(c >= "99204" for c in em_codes):
        score += 14
        risks.append("HIGH RISK: high-complexity E/M without active diagnosis codes on claim.")
        recs.append("Strengthen clinical documentation and diagnosis linkage.")

    if score >= 38:
        level = "High"
    elif score >= 20:
        level = "Medium"
    else:
        level = "Low"

    if not risks:
        risks.append("Properly matched coding — continue routine documentation review.")
    if not recs:
        recs.append("Maintain complete chart and timely filing.")

    return {
        "riskLevel": level,
        "riskScore": min(100, score),
        "risks": risks[:12],
        "recommendations": recs[:12],
    }


def run_intelligence_analysis(payload: Dict[str, Any]) -> Dict[str, Any]:
    """Single payload: cpt_codes / cptCodes; icd10_codes (canonical) / icd_codes / icdCodes / diagnosis_code; carcCodes; planType."""
    cpt = payload.get("cpt_codes") or payload.get("cptCodes") or ""
    icd = (
        payload.get("icd10_codes")
        or payload.get("icd_codes")
        or payload.get("icdCodes")
        or payload.get("diagnosis_code")
        or ""
    )

    carc_list = payload.get("carcCodes") or payload.get("carc_codes")
    if carc_list is None and payload.get("carc_codes"):
        carc_list = payload.get("carc_codes")

    denial = payload.get("denial_codes") or payload.get("denial_code") or ""
    denial += " " + str(payload.get("carc_codes") or "") + " " + str(payload.get("rarc_codes") or "")
    if isinstance(carc_list, list):
        denial += " " + " ".join(str(x) for x in carc_list)

    coding = validateCoding({"cptCodes": cpt, "icdCodes": icd})
    modifiers = suggestModifiers(
        {
            "cptCodes": cpt,
            "carcCodes": carc_list if carc_list is not None else payload.get("carc_codes"),
            "context": {
                "denial_reason": payload.get("denial_reason"),
                "global_period_unrelated": payload.get("global_period_unrelated"),
            },
        },
        denial,
        None,
    )
    risk = detectDenialRisk(payload)
    return {
        "coding": coding,
        "modifiers": modifiers,
        "risk": risk,
    }


def validateCoding_export(cpt_codes: Any, icd_codes: Any) -> Dict[str, Any]:
    return validateCoding(cpt_codes, icd_codes)

File: lib/intelligence/coding-intelligence.ts
/**
 * Port of backend/coding_intelligence.py — coding validation, modifiers, denial risk.
 * Keep behavior aligned with Flask for intake /api/intelligence/analyze.
 */

export type JsonRecord = Record<string, unknown>;

const EM_RANGE = /^99[2-4]\d{2}$/;
const HIGH_OFFICE_EM = /^992(0[4-5]|1[3-5])$/;
const PROCEDURE_SURGICAL = /^(1\d{4}|2\d{4}|3\d{4}|4\d{4}|5\d{4}|6\d{4})$/;
const IMAGING = /^(7\d{4})$/;

function splitCodes(raw: unknown): string[] {
  if (raw == null) return [];
  const parts = Array.isArray(raw) ? raw : String(raw).trim().split(/[\s,;]+/);
  const out: string[] = [];
  for (const p0 of parts) {
    let p = String(p0).trim().toUpperCase().replace(/\./g, "");
    if (!p) continue;
    const base = p.match(/^(\d{5})/);
    if (base) out.push(base[1]!);
    else if (/^\d{5}$/.test(p)) out.push(p);
  }
  return [...new Set(out)].slice(0, 40);
}

function splitIcd(raw: unknown): string[] {
  if (raw == null) return [];
  const parts = Array.isArray(raw) ? raw : String(raw).trim().split(/[\s,;]+/);
  const out: string[] = [];
  for (const p0 of parts) {
    const p = String(p0).replace(/[^A-Z0-9.]/gi, "").toUpperCase();
    if (p.length >= 3) out.push(p);
  }
  return [...new Set(out)].slice(0, 24);
}

function icdSpecificity(icd: string): "unspecified" | "general_z" | "specific" | "moderate" {
  const core = icd.replace(/\./g, "");
  if (core.length <= 3) return "unspecified";
  if (core.startsWith("Z") && core.length <= 5) return "general_z";
  if (/^[A-Z]\d{2}\d{3}$/.test(core) || /^[A-Z]\d{2}\.\d/.test(icd)) return "specific";
  return "moderate";
}

function normalizeValidatePair(cpt_codes: unknown, icd_codes: unknown, kwargs: JsonRecord): [unknown, unknown] {
  if (cpt_codes != null && typeof cpt_codes === "object" && !Array.isArray(cpt_codes)) {
    const d = cpt_codes as JsonRecord;
    return [
      d.cptCodes ?? d.cpt_codes,
      d.icdCodes ?? d.icd10_codes ?? d.icd_codes,
    ];
  }
  const cpt = kwargs.cptCodes != null ? kwargs.cptCodes : cpt_codes;
  const icd = kwargs.icdCodes != null ? kwargs.icdCodes : icd_codes;
  return [cpt, icd];
}

export function validateCoding(cpt_codes?: unknown, icd_codes?: unknown, kwargs: JsonRecord = {}): JsonRecord {
  const [cpt_in, icd_in] = normalizeValidatePair(cpt_codes, icd_codes, kwargs);
  const cpts = splitCodes(cpt_in);
  const icds = splitIcd(icd_in);
  const issues: string[] = [];
  const suggestions: string[] = [];
  const weakIcd: string[] = [];

  if (!cpts.length) {
    issues.push("No CPT/HCPCS codes provided — cannot validate medical necessity pairing.");
    suggestions.push("Enter at least one procedure or visit code.");
  }
  if (!icds.length) {
    issues.push("No ICD-10 diagnosis codes provided — missing diagnosis linkage.");
    suggestions.push("Add diagnosis codes that support medical necessity for each service line.");
  }

  const emCodes = cpts.filter((c) => EM_RANGE.test(c));
  const procCodes = cpts.filter((c) => PROCEDURE_SURGICAL.test(c));
  const imgCodes = cpts.filter((c) => IMAGING.test(c));

  for (const icd of icds) {
    const spec = icdSpecificity(icd);
    if (spec === "unspecified" || spec === "general_z") weakIcd.push(icd);
  }

  for (const em of emCodes) {
    if (icds.length) {
      const onlyZ = icds.every((i) => i.startsWith("Z"));
      if (onlyZ && HIGH_OFFICE_EM.test(em)) {
        issues.push(
          `ICD-10 pattern may not support CPT ${em} (high-level E/M with only Z-series diagnoses).`
        );
        suggestions.push("Add active diagnoses documented for this visit or align E/M level with payer policy.");
      }
      if (onlyZ && em >= "99204" && !HIGH_OFFICE_EM.test(em)) {
        issues.push(
          `ICD-10 may not fully support CPT ${em} for medical necessity if only screening codes apply.`
        );
        suggestions.push("Ensure visit diagnosis matches the work performed.");
      }
    }
  }

  if (emCodes.length && icds.length) {
    const onlyZScreening = icds.every((i) => i.startsWith("Z"));
    if (onlyZScreening && emCodes.some((c) => c >= "99204")) {
      if (!issues.some((x) => x.includes("ICD-10 pattern may not support CPT"))) {
        issues.push("Diagnosis is too general or screening-only for the service complexity reported.");
        suggestions.push("Use diagnoses that support medical necessity for the E/M level billed.");
      }
    }
  }

  if (emCodes.length && procCodes.length) {
    issues.push("E/M and procedure on the same encounter often require modifier -25 on the E/M when distinct.");
    suggestions.push("Confirm separate E/M service from procedure; consider modifier -25 if payer rules allow.");
  }

  for (const icd of icds) {
    if (icdSpecificity(icd) === "unspecified") {
      issues.push(`ICD-10 ${icd} appears truncated or unspecified — may lack specificity for the service.`);
      suggestions.push("Use the most specific billable ICD-10 supported by documentation.");
    }
  }

  if (procCodes.length && !emCodes.length && icds.length === 1) {
    const s = icdSpecificity(icds[0]!);
    if (s === "general_z" || s === "unspecified") {
      issues.push("Procedure coding with a single broad or Z-code diagnosis may indicate incomplete linkage.");
      suggestions.push("Link complications, anatomy, or active conditions as documented.");
    }
  }

  if (imgCodes.length && icds.length) {
    if (icds.every((i) => ["unspecified", "general_z", "moderate"].includes(icdSpecificity(i)))) {
      issues.push("Imaging CPT with broadly stated diagnoses — payer may question medical necessity specificity.");
      suggestions.push("Tighten ICD-10 to findings or symptoms that justify the study.");
    }
  }

  if (imgCodes.length) {
    issues.push("Imaging CPT codes often require prior authorization by payer and policy.");
    suggestions.push("Verify authorization on file for the DOS and modality.");
  }

  const valid = issues.length === 0;
  return {
    valid,
    issues: issues.slice(0, 14),
    suggestions: suggestions.slice(0, 14),
    weakIcdCodes: [...new Set(weakIcd)].slice(0, 24),
    cpt_parsed: cpts,
    icd_parsed: icds,
  };
}

function parseCarcFromText(blob: string): number[] {
  const out: number[] = [];
  const re = /(?:CO|PR|PI|OA)?[-\s]?(\d{1,3})/g;
  let m: RegExpExecArray | null;
  const u = (blob || "").toUpperCase();
  while ((m = re.exec(u)) != null) {
    const n = parseInt(m[1]!, 10);
    if (n > 0 && n <= 999) out.push(n);
  }
  return [...new Set(out)].slice(0, 24);
}

function carcFromAny(carc_codes: unknown, denialBlob: string): number[] {
  let nums = parseCarcFromText(denialBlob);
  if (Array.isArray(carc_codes)) {
    for (const x of carc_codes) nums = nums.concat(parseCarcFromText(String(x)));
  } else if (carc_codes) {
    nums = nums.concat(parseCarcFromText(String(carc_codes)));
  }
  return [...new Set(nums)].slice(0, 24);
}

function normalizeSuggestArgs(
  cpt_codes: unknown,
  denial_codes: unknown,
  context: JsonRecord | null | undefined
): [string[], number[], JsonRecord, string] {
  let ctx: JsonRecord = { ...(context || {}) };
  let cpts: string[] = [];
  let blob = "";
  let carcs: number[] = [];

  if (cpt_codes != null && typeof cpt_codes === "object" && !Array.isArray(cpt_codes)) {
    const d = cpt_codes as JsonRecord;
    const cptRaw = d.cptCodes ?? d.cpt_codes;
    const carcRaw = d.carcCodes ?? d.carc_codes;
    ctx = { ...ctx, ...((d.context as JsonRecord) || {}) };
    cpts = splitCodes(cptRaw);
    blob = Array.isArray(carcRaw) ? carcRaw.map((x) => String(x)).join(" ") : String(carcRaw || "");
    carcs = carcFromAny(carcRaw, blob);
    if (denial_codes != null && denial_codes !== "") {
      const extra = Array.isArray(denial_codes) ? denial_codes.map((x) => String(x)).join(" ") : String(denial_codes);
      blob = `${blob} ${extra}`.trim();
    }
  } else {
    cpts = splitCodes(cpt_codes);
    blob = Array.isArray(denial_codes) ? denial_codes.map((x) => String(x)).join(" ") : String(denial_codes || "");
    carcs = carcFromAny(null, blob);
  }
  blob = `${blob} ${String(ctx.denial_reason || "")}`.trim();
  carcs = [...new Set([...carcs, ...parseCarcFromText(blob)])].slice(0, 24);
  return [cpts, carcs, ctx, blob];
}

export function suggestModifiers(
  cpt_codes?: unknown,
  denial_codes?: unknown,
  context?: JsonRecord | null
): JsonRecord {
  const [cpts, carcs, ctx, blob] = normalizeSuggestArgs(cpt_codes, denial_codes, context);
  const recommended: string[] = [];
  const reasoningLines: string[] = [];
  const U = blob.toUpperCase();

  const emCodes = cpts.filter((c) => EM_RANGE.test(c));
  const procCodes = cpts.filter((c) => PROCEDURE_SURGICAL.test(c));
  const imgCodes = cpts.filter((c) => IMAGING.test(c));
  const procLike = procCodes.concat(imgCodes);

  if (emCodes.length && procLike.length) {
    recommended.push("-25");
    reasoningLines.push("Separate E/M service performed with procedure — distinct service when documented.");
  }
  if (carcs.includes(97) || U.includes("BUNDL")) {
    if (!recommended.includes("-59") && !recommended.includes("-XS")) recommended.push("-59");
    reasoningLines.push("CARC 97 (bundling / inclusive payment) — distinct procedural service may require -59 or -XS.");
  }
  if (carcs.includes(4) || U.includes("MODIFIER")) {
    if (!recommended.includes("-59")) recommended.push("-59");
    reasoningLines.push("CARC 4 or modifier-related denial — review NCCI and append -59 or appropriate modifier when distinct.");
  }
  if (ctx.global_period_unrelated || U.includes("GLOBAL")) {
    recommended.push("-24");
    reasoningLines.push("Unrelated E/M during postoperative global period — use -24 when policy criteria are met.");
  }
  if (U.includes("REPEAT") || carcs.includes(76)) {
    recommended.push("-76");
    reasoningLines.push("Repeat procedure by same physician — -76 when the same procedure is repeated.");
  }
  if (U.includes("DIFFERENT") && U.includes("PHYSICIAN")) {
    recommended.push("-77");
    reasoningLines.push("Repeat procedure, different physician — consider -77.");
  }
  const nProc = cpts.filter((c) => PROCEDURE_SURGICAL.test(c) || IMAGING.test(c)).length;
  if (nProc >= 2) {
    if (!recommended.includes("-59") && !recommended.includes("-XS")) recommended.push("-59");
    reasoningLines.push("Multiple procedures — evaluate distinct services; -59 or -XS when separately reportable.");
  }

  const seen = new Set<string>();
  const mods: string[] = [];
  for (const m of recommended) {
    if (!seen.has(m)) {
      seen.add(m);
      mods.push(m);
    }
  }
  if (!reasoningLines.length) {
    reasoningLines.push("No automatic modifier rules triggered — verify payer policy and NCCI edits.");
  }
  return {
    recommendedModifiers: mods.slice(0, 8),
    reasoning: reasoningLines.slice(0, 12),
    matched_carcs: carcs,
  };
}

export function detectDenialRisk(inputData: JsonRecord): JsonRecord {
  const risks: string[] = [];
  const recs: string[] = [];
  let score = 0;

  const cpts = splitCodes(inputData.cpt_codes ?? inputData.cptCodes);
  const icds = splitIcd(
    inputData.icd10_codes ?? inputData.icd_codes ?? inputData.diagnosis_code ?? inputData.icdCodes
  );
  const mods = String(inputData.modifiers || "").toUpperCase();
  const payer = String(inputData.payer || "");
  const plan = String(inputData.planType || inputData.plan_type || "");

  let carcBlob = [inputData.denial_code, inputData.carc_codes, inputData.carcCodes].map((x) => String(x ?? "")).join(" ");
  if (Array.isArray(inputData.carcCodes)) {
    carcBlob += " " + inputData.carcCodes.map((x) => String(x)).join(" ");
  }
  const denialBlob = [carcBlob, inputData.denial_reason, inputData.rarc_codes].map((x) => String(x ?? "")).join(" ");
  const carcNums = parseCarcFromText(denialBlob);

  if (!cpts.length) {
    score += 15;
    risks.push("Missing procedure codes — high scrubber / medical necessity exposure.");
    recs.push("Enter all line-item CPT/HCPCS codes.");
  }
  if (!icds.length) {
    score += 22;
    risks.push("Missing diagnosis codes — high likelihood of medical necessity denial.");
    recs.push("Add ICD-10 codes tied to documentation.");
  }
  for (const icd of icds) {
    if ((icdSpecificity(icd) === "unspecified" || icdSpecificity(icd) === "general_z") && cpts.length) {
      score += 12;
      risks.push("Weak ICD-10 specificity relative to billed services.");
      recs.push("Use specific, billable diagnoses supported by the record.");
      break;
    }
  }

  const emCodes = cpts.filter((c) => EM_RANGE.test(c));
  const procCodes = cpts.filter((c) => PROCEDURE_SURGICAL.test(c));
  if (emCodes.length && procCodes.length && !mods.includes("-25") && !mods.includes("25")) {
    score += 18;
    risks.push("HIGH RISK: E/M plus procedure same day without modifier -25 — likely bundling or denial.");
    recs.push("Add modifier -25 to the E/M when separately identifiable.");
  }
  if (carcNums.includes(197) || carcNums.includes(16)) {
    score += 14;
    risks.push("Authorization / precertification cited on claim (CARC pattern).");
    recs.push("Secure authorization or retro-auth documentation before resubmission.");
  }
  if (carcNums.includes(252)) {
    score += 8;
    risks.push("Additional documentation or authorization gap (CARC 252 pattern).");
    recs.push("Submit complete clinical packet or auth proof.");
  }
  if (cpts.some((c) => IMAGING.test(c))) {
    score += 8;
    risks.push("Authorization-sensitive CPT: advanced imaging often requires precert.");
    recs.push("Confirm precertification and medical necessity narrative.");
  }
  if (plan.toUpperCase() === "MEDICAID" || plan.toUpperCase() === "MEDICARE") {
    score += 5;
    risks.push(`${plan} plans often apply strict LCD / NCD medical necessity rules.`);
    recs.push("Align diagnoses and modifiers with program coverage policies.");
  }
  if (payer.toUpperCase().includes("UNITED") || payer.toUpperCase().includes("UHC")) {
    score += 6;
    risks.push("Known payer pattern: strict policy-driven medical necessity reviews.");
    recs.push("Cite applicable clinical policy in documentation.");
  }
  if (emCodes.length && icds.length && icds.every((i) => i.startsWith("Z")) && emCodes.some((c) => c >= "99204")) {
    score += 14;
    risks.push("HIGH RISK: high-complexity E/M without active diagnosis codes on claim.");
    recs.push("Strengthen clinical documentation and diagnosis linkage.");
  }

  if (!risks.length) {
    risks.push("Properly matched coding — continue routine documentation review.");
  }
  if (!recs.length) {
    recs.push("Maintain complete chart and timely filing.");
  }
  const level = score >= 38 ? "High" : score >= 20 ? "Medium" : "Low";
  return {
    riskLevel: level,
    riskScore: Math.min(100, score),
    risks: risks.slice(0, 12),
    recommendations: recs.slice(0, 12),
  };
}

export function runIntelligenceAnalysis(payload: JsonRecord): JsonRecord {
  const cpt = payload.cpt_codes ?? payload.cptCodes ?? "";
  const icd =
    payload.icd10_codes ?? payload.icd_codes ?? payload.icdCodes ?? payload.diagnosis_code ?? "";
  let carcList: unknown = payload.carcCodes ?? payload.carc_codes;
  if (carcList == null && payload.carc_codes) carcList = payload.carc_codes;
  let denial = String(payload.denial_codes || payload.denial_code || "");
  denial += ` ${String(payload.carc_codes || "")} ${String(payload.rarc_codes || "")}`;
  if (Array.isArray(carcList)) {
    denial += " " + carcList.map((x) => String(x)).join(" ");
  }
  const coding = validateCoding({ cptCodes: cpt, icdCodes: icd });
  const modifiers = suggestModifiers(
    {
      cptCodes: cpt,
      carcCodes: carcList != null ? carcList : payload.carc_codes,
      context: {
        denial_reason: payload.denial_reason,
        global_period_unrelated: payload.global_period_unrelated,
      },
    },
    denial,
    undefined
  );
  const risk = detectDenialRisk(payload);
  return { coding, modifiers, risk };
}
